medcat/relation_extraction.py

```diff
@@ -0,0 +1,293 @@
+       self.batch_size = config.train["batch_size"]
+       self.n_classes = config.model["nclasses"]
+
+       self.is_cuda_available = torch.cuda.is_available()
+
+       self.device = torch.device("cuda:0" if self.is_cuda_available else "cpu")
+       self.hf_model_name = "bert-large-uncased"
+
+       self.model_config = BertConfig.from_pretrained(self.hf_model_name) 
+
+       if self.is_cuda_available:
+           self.model = self.model.to(self.device)
+
+       if self.tokenizer is None:
+            tokenizer_path = os.path.join(re_model_path, "BERT_tokenizer_relation_extraction")
+            if os.path.exists(tokenizer_path):
+                self.tokenizer = TokenizerWrapperBERT.load(tokenizer_path)
+            else:
+                self.tokenizer = TokenizerWrapperBERT(AutoTokenizer.from_pretrained(pretrained_model_name_or_path="bert-large-uncased"),
+                                                      max_seq_length=self.model_config.max_position_embeddings) 
+                create_tokenizer_pretrain(self.tokenizer)
+    
+       self.model_config.vocab_size = len(self.tokenizer.hf_tokenizers)
+
+       self.model = BertModel_RelationExtraction.from_pretrained(pretrained_model_name_or_path=self.hf_model_name,
+                                                                 model_size=self.hf_model_name,
+                                                                 config=self.model_config,
+                                                                 task=task,
+                                                                 n_classes=self.n_classes)  
+
+       self.model.resize_token_embeddings(self.model_config.vocab_size)
+       
+       unfrozen_layers = ["classifier", "pooler", "encoder.layer.11", \
+                          "classification_layer", "blanks_linear", "lm_linear", "cls"]
+
+       for name, param in self.model.named_parameters():
+        if not any([layer in name for layer in unfrozen_layers]):
+            param.requires_grad = False
+        else:
+            param.requires_grad = True
+        
+       self.pad_id = self.tokenizer.hf_tokenizers.pad_token_id
+       self.padding_seq = Pad_Sequence(seq_pad_value=self.pad_id,\
+                       label_pad_value=self.pad_id,\
+                       label2_pad_value=-1)
+
+    def create_test_train_datasets(self, data):
+        train_data, test_data = {}, {}
+        train_data["output_relations"], test_data["output_relations"] = split_list_train_test(data["output_relations"],
+                                    test_size=self.config.train["test_size"], shuffle=False)
+        for k,v in data.items():
+            if k != "output_relations":
+                train_data[k] = v
+                test_data[k] = v
+
+        return train_data, test_data
+
+    def train(self, export_data_path = "", csv_path = "", docs = None, checkpoint_path="./", num_epoch=1, gradient_acc_steps=1, multistep_lr_gamma=0.8, max_grad_norm=1.0):

```

Comment:
Why are the training params specified here, but also specified in the  config_re.ConfigRE?  num_epoch, gradient_acc_steps etc.?

Reply:
I've now added usage from config instead.